File: fushi/test/anki/exported_glossary_anchor_deeplink_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:fushi/src/lookup/lookup_deep_link.dart';

/// BUG-1666：制卡后卡片上「词典释义里指向另一个单词」的链接点开跳向
/// `http://127.0.0.1:<端口>`（Anki 桌面/AnkiDroid 的本地媒体服务器 base URL 把
/// 释义里保留的 `entry://` / 相对路径交叉引用解析成了死本地页）。
///
/// 修复分三段，三层守护对应：
/// ① 行为级——node 真执行 popup.js 的 `rewriteExportedGlossaryAnchors`，断言
///    entry://、相对路径改写成 `fushi://lookup?word=<词>` 深链、外链/`#` 保留、
///    sound:// 与无文本锚点去 href；并源级断言两个导出构建器
///    （constructGlossaryHtml / constructSingleGlossaryHtml）都在序列化前调用它。
///    无 node 时 skip。
/// ② 三镜像逐字节一致由 `browser_extension_popup_parity_guard_test.dart` 守。
/// ③ 深链消费端——`lookupWordFromDeepLink` 纯函数（Windows 冷启动 argv /
///    单实例 WM_COPYDATA 转交两条路共用）在本文件直接单测；Android 端同参数由
///    Kotlin `extractProcessText` 解析（scheme 须含 "fushi"，见源级守卫）。
void main() {
  test(
    'BUG-1666: exported glossary cross-references become fushi://lookup deep '
    'links (executes popup.js via node)',
    () async {
      final String? nodeExe = _resolveNode();
      if (nodeExe == null) {
        markTestSkipped(
            'node not found on PATH; skipping JS behavior execution');
        return;
      }

      final File jsTest =
          File('test/anki/exported_glossary_anchor_deeplink_test.js');
      expect(jsTest.existsSync(), isTrue,
          reason: 'behavior harness ${jsTest.path} must exist');

      final ProcessResult result = await Process.run(
        nodeExe,
        <String>[jsTest.path],
        workingDirectory: Directory.current.path,
      );

      expect(
        result.exitCode,
        0,
        reason: 'BUG-1666 exported anchor deep-link behavior test failed.\n'
            'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
      );
      expect(result.stdout.toString(), contains('all assertions passed'),
          reason: 'behavior harness must reach its success marker');
    },
  );

  group('lookupWordFromDeepLink', () {
    test('parses fushi://lookup?word=… (percent-decoded, trimmed)', () {
      expect(lookupWordFromDeepLink('fushi://lookup?word=belong'), 'belong');
      expect(
        lookupWordFromDeepLink(
            'fushi://lookup?word=%E9%A3%9F%E3%81%B9%E3%82%8B'),
        '食べる',
      );
      expect(lookupWordFromDeepLink(' fushi://lookup?word=%20belong%20 '),
          'belong');
    });

    test('accepts the legacy hibiki scheme', () {
      expect(lookupWordFromDeepLink('hibiki://lookup?word=belong'), 'belong');
    });

    test('rejects non-lookup URLs and empty words', () {
      expect(lookupWordFromDeepLink('fushi://auth/onedrive?code=x'), isNull);
      expect(lookupWordFromDeepLink('fushi://lookup'), isNull);
      expect(lookupWordFromDeepLink('fushi://lookup?word='), isNull);
      expect(lookupWordFromDeepLink('fushi://lookup?word=%20'), isNull);
      expect(
          lookupWordFromDeepLink('https://example.com/?word=belong'), isNull);
      expect(lookupWordFromDeepLink(r'D:\video\ep01.mkv'), isNull);
      expect(lookupWordFromDeepLink(''), isNull);
    });
  });

  test(
      'Android :popup activities accept the fushi scheme for lookup deep links '
      '(source guard)', () {
    const List<String> activities = <String>[
      'android/app/src/main/java/app/fushi/reader/PopupDictFlutterActivity.kt',
      'android/app/src/main/java/app/fushi/reader/PopupDictActivity.kt',
    ];
    for (final String path in activities) {
      final File file = File(path);
      expect(file.existsSync(), isTrue, reason: '$path must exist');
      final String source = file.readAsStringSync();
      // 断言字面量写进注释，变异测试时能定位：manifest 注册的 scheme 是
      // "fushi"，Kotlin 侧曾只认 "hibiki"（改名残留）导致深链开出空弹窗。
      expect(source, contains('uri.scheme == "fushi"'),
          reason: '$path must accept the fushi scheme (manifest registers it)');
    }
  });
}

String? _resolveNode() {
  final String exe = Platform.isWindows ? 'node.exe' : 'node';
  final String pathEnv = Platform.environment['PATH'] ?? '';
  final String separator = Platform.isWindows ? ';' : ':';
  for (final String dir in pathEnv.split(separator)) {
    if (dir.isEmpty) continue;
    final File candidate = File('$dir${Platform.pathSeparator}$exe');
    if (candidate.existsSync()) return candidate.path;
  }
  return null;
}

File: fushi/test/anki/exported_glossary_anchor_deeplink_test.js
// BUG-1666 行为守卫：制卡导出的释义 HTML 里，词典内部交叉引用锚点必须改写成
// fushi://lookup?word=<可见词头> 深链，而不是保留 entry:// / 相对路径原样。
//
// 症状：Anki（桌面 + AnkiDroid）用本地媒体服务器（http://127.0.0.1:<随机端口>）
// 作卡片 WebView 的 base URL，释义里指向「另一个单词」的相对链接点下去就解析成
// 127.0.0.1 的死页。弹窗内这类点击被 handleGlossaryAnchorClick（BUG-767）拦截并按
// 可见词头重查；卡片上没有拦截器，只能把同一语义烤进导出字节。
//
// 判据（rewriteExportedGlossaryAnchors，popup.js）：
//   - entry:// 与相对路径 → fushi://lookup?word=<encodeURIComponent(textContent)>
//   - http(s):// 外链保留
//   - # 片段保留（不离开卡片，无 127 危害）
//   - sound:/image:/dictmedia:（字节未导出）→ 去掉 href
//   - 无可见文本的内部锚点 → 去掉 href
//
// 运行：node fushi/test/anki/exported_glossary_anchor_deeplink_test.js
// 由同名 .dart wrapper 通过 Process.run('node', ...) 驱动（无 node 时 skip）。

const assert = require('node:assert');
const fs = require('node:fs');
const path = require('node:path');
const vm = require('node:vm');

const popupSrc = fs.readFileSync(
  path.resolve(__dirname, '..', '..', 'assets', 'popup', 'popup.js'),
  'utf8',
);

// ---- 提取被测函数（不执行整个 popup.js：它依赖真实 DOM/window） ------------
const fnMatch = popupSrc.match(
  /function rewriteExportedGlossaryAnchors\(root\) \{[\s\S]*?\n\}/,
);
assert.ok(fnMatch, 'popup.js must define rewriteExportedGlossaryAnchors(root)');

const context = { console };
vm.createContext(context);
vm.runInContext(`${fnMatch[0]}; this.__fn = rewriteExportedGlossaryAnchors;`, context);
const rewriteExportedGlossaryAnchors = context.__fn;

// ---- 两个导出构建器都必须在序列化前调用它（调用点源级判据） ----------------
const callSites = popupSrc.match(/rewriteExportedGlossaryAnchors\(tempDiv\);/g) || [];
assert.ok(
  callSites.length >= 2,
  `both export builders (constructGlossaryHtml / constructSingleGlossaryHtml) must call ` +
  `rewriteExportedGlossaryAnchors(tempDiv); found ${callSites.length} call site(s)`,
);

// ---- 最小 fake DOM：只需 querySelectorAll('a[href]') + 锚点三方法 ----------
function makeAnchor(href, text) {
  const attrs = { href };
  return {
    textContent: text,
    getAttribute: (name) => (name in attrs ? attrs[name] : null),
    setAttribute: (name, value) => { attrs[name] = value; },
    removeAttribute: (name) => { delete attrs[name]; },
    get href() { return attrs.href; },
    hasHref() { return 'href' in attrs; },
  };
}

function makeRoot(anchors) {
  return {
    querySelectorAll: (selector) => {
      assert.strictEqual(selector, 'a[href]');
      return anchors;
    },
  };
}

// 场景 A：MDX entry:// 交叉引用 → fushi 深链（按可见词头，percent-encode）。
{
  const a = makeAnchor('entry://belong', 'belong');
  rewriteExportedGlossaryAnchors(makeRoot([a]));
  assert.strictEqual(a.href, 'fushi://lookup?word=belong', 'entry:// must become a fushi deep link');
}

// 场景 B：相对路径（Anki 卡片里会解析到 127.0.0.1 本地服务器）→ fushi 深链。
{
  const a = makeAnchor('/belong_1', 'belong');
  rewriteExportedGlossaryAnchors(makeRoot([a]));
  assert.strictEqual(a.href, 'fushi://lookup?word=belong', 'relative href must become a fushi deep link');
}

// 场景 C：可见词头含非 ASCII / 空白 → encodeURIComponent + trim。
{
  const a = makeAnchor('entry://食べる（たべる）', ' 食べる ');
  rewriteExportedGlossaryAnchors(makeRoot([a]));
  assert.strictEqual(
    a.href,
    `fushi://lookup?word=${encodeURIComponent('食べる')}`,
    'visible headword must be trimmed and percent-encoded',
  );
}

// 场景 D：真外链保留原样。
{
  const a = makeAnchor('https://example.com/belong', 'belong');
  rewriteExportedGlossaryAnchors(makeRoot([a]));
  assert.strictEqual(a.href, 'https://example.com/belong', 'external http(s) links must be kept');
}

// 场景 E：# 片段保留（卡内跳转无 127 危害）。
{
  const a = makeAnchor('#idiom-1', 'Idioms');
  rewriteExportedGlossaryAnchors(makeRoot([a]));
  assert.strictEqual(a.href, '#idiom-1', '# fragments must be kept');
}

// 场景 F：sound:// 发音锚点（音频字节未随卡导出）→ 去 href，不指向任何死地址。
{
  const a = makeAnchor('sound://media/belong.mp3', '🔊');
  rewriteExportedGlossaryAnchors(makeRoot([a]));
  assert.strictEqual(a.hasHref(), false, 'sound:// anchors must lose their href');
}

// 场景 G：无可见文本的内部锚点（图标图等）→ 去 href。
{
  const a = makeAnchor('entry://belong', '  ');
  rewriteExportedGlossaryAnchors(makeRoot([a]));
  assert.strictEqual(a.hasHref(), false, 'internal anchors without visible text must lose their href');
}

console.log('all assertions passed');
